# Ames Housing Price Prediction

**Team:** The Outliers  
**Course:** Advanced Apex Project 1  
**Institution:** BITS Pilani

## Project Overview

This analysis develops machine learning models to predict residential property sale prices in Ames, Iowa. Using a dataset of 2,930 home sales with 82 features, we perform comprehensive data exploration, feature engineering, and model development to achieve accurate price predictions.

## Table of Contents

1. [Phase 1: Data Acquisition](#phase1)
2. [Phase 2A: Data Preprocessing & Exploratory Analysis](#phase2a)
3. [Phase 2B: Feature Engineering](#phase2b)
4. [Phase 3: Model Development & Evaluation](#phase3)

---

<a id='phase1'></a>
## Phase 1: Data Acquisition

**Objective:** Load and perform initial inspection of the Ames Housing dataset.

### 1.1 Import Dependencies

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)
sns.set_style('whitegrid')

print("Libraries imported successfully.")

Libraries imported successfully.


### 1.2 Dataset Import

In [2]:
df = pd.read_csv("../data/AmesHousing.csv")

print(f"Dataset shape: {df.shape}")
print(f"Total records: {df.shape[0]:,}")
print(f"Total features: {df.shape[1]}")
print("\nFirst 5 rows:")
df.head()

Dataset shape: (2930, 82)
Total records: 2,930
Total features: 82

First 5 rows:


Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL        141.00     31770   Pave   
1      2  526350040           20        RH         80.00     11622   Pave   
2      3  526351010           20        RL         81.00     14267   Pave   
3      4  526353030           20        RL         93.00     11160   Pave   
4      5  527105010           60        RL         74.00     13830   Pave   

  Alley Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0   NaN       IR1          Lvl    AllPub     Corner        Gtl        NAmes   
1   NaN       Reg          Lvl    AllPub     Inside        Gtl        NAmes   
2   NaN       IR1          Lvl    AllPub     Corner        Gtl        NAmes   
3   NaN       Reg          Lvl    AllPub     Corner        Gtl        NAmes   
4   NaN       IR1          Lvl    AllPub     Inside        Gtl      Gilbert   

  Condition 1 Condition 2 Bldg Type House Style  Overall Qual  Overall Cond  \
0        Norm        Norm      1Fam      1Story             6             5   
1       Feedr        Norm      1Fam      1Story             5             6   
2        Norm        Norm      1Fam      1Story             6             6   
3        Norm        Norm      1Fam      1Story             7             5   
4        Norm        Norm      1Fam      2Story             5             5   

   Year Built  Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd  \
0        1960            1960        Hip   CompShg      BrkFace      Plywood   
1        1961            1961      Gable   CompShg      VinylSd      VinylSd   
2        1958            1958        Hip   CompShg      Wd Sdng      Wd Sdng   
3        1968            1968        Hip   CompShg      BrkFace      BrkFace   
4        1997            1998      Gable   CompShg      VinylSd      VinylSd   

  Mas Vnr Type  Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual  \
0        Stone        112.00         TA         TA     CBlock        TA   
1          NaN          0.00         TA         TA     CBlock        TA   
2      BrkFace        108.00         TA         TA     CBlock        TA   
3          NaN          0.00         Gd         TA     CBlock        TA   
4          NaN          0.00         TA         TA      PConc        Gd   

  Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  \
0        Gd            Gd            BLQ        639.00            Unf   
1        TA            No            Rec        468.00            LwQ   
2        TA            No            ALQ        923.00            Unf   
3        TA            No            ALQ       1065.00            Unf   
4        TA            No            GLQ        791.00            Unf   

   BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air  \
0          0.00       441.00        1080.00    GasA         Fa           Y   
1        144.00       270.00         882.00    GasA         TA           Y   
2          0.00       406.00        1329.00    GasA         TA           Y   
3          0.00      1045.00        2110.00    GasA         Ex           Y   
4          0.00       137.00         928.00    GasA         Gd           Y   

  Electrical  1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  \
0      SBrkr        1656           0                0         1656   
1      SBrkr         896           0                0          896   
2      SBrkr        1329           0                0         1329   
3      SBrkr        2110           0                0         2110   
4      SBrkr         928         701                0         1629   

   Bsmt Full Bath  Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  \
0            1.00            0.00          1          0              3   
1            0.00            0.00          1          0              2   
2            0.00            0.00          1          1              3   
3            1.00            0.00

### 1.3 Data Schema Verification

In [3]:
print("\nColumn Names:")
print(df.columns.tolist())

print("\nData Types:")
print(df.dtypes.value_counts())

print("\nBasic Info:")
df.info()


Column Names:
['Order', 'PID', 'MS SubClass', 'MS Zoning', 'Lot Frontage', 'Lot Area', 'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities', 'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1', 'Condition 2', 'Bldg Type', 'House Style', 'Overall Qual', 'Overall Cond', 'Year Built', 'Year Remod/Add', 'Roof Style', 'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type', 'Mas Vnr Area', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin SF 1', 'BsmtFin Type 2', 'BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF', 'Heating', 'Heating QC', 'Central Air', 'Electrical', '1st Flr SF', '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Bsmt Full Bath', 'Bsmt Half Bath', 'Full Bath', 'Half Bath', 'Bedroom AbvGr', 'Kitchen AbvGr', 'Kitchen Qual', 'TotRms AbvGrd', 'Functional', 'Fireplaces', 'Fireplace Qu', 'Garage Type', 'Garage Yr Blt', 'Garage Finish', 'Garage Cars', 'Garage Area', 'Garage Qual', 'Garage Cond', 'Paved Driv

---

<a id='phase2a'></a>
## Phase 2A: Data Preprocessing & Exploratory Analysis

**Objective:** Understand data quality, distributions, and relationships.

### 2.1 Missing Value Analysis

We systematically identify and quantify missing data across all features.

In [4]:
missing_counts = df.isnull().sum()
missing_pct = (missing_counts / len(df)) * 100

missing_df = pd.DataFrame({
    'Count': missing_counts[missing_counts > 0],
    'Percentage': missing_pct[missing_counts > 0]
}).sort_values('Percentage', ascending=False)

print(f"Features with missing data: {len(missing_df)}")
print("\nTop 15 features with most missing data:")
print(missing_df.head(15))

Features with missing data: 27

Top 15 features with most missing data:
                Count  Percentage
Pool QC          2917       99.56
Misc Feature     2824       96.38
Alley            2732       93.24
Fence            2358       80.48
Mas Vnr Type     1775       60.58
Fireplace Qu     1422       48.53
Lot Frontage      490       16.72
Garage Qual       159        5.43
Garage Cond       159        5.43
Garage Yr Blt     159        5.43
Garage Finish     159        5.43
Garage Type       157        5.36
Bsmt Exposure      83        2.83
BsmtFin Type 2     81        2.76
Bsmt Cond          80        2.73


### 2.2 Missing Value Visualization

Visual representations help identify patterns in missing data.

In [ ]:
plt.figure(figsize=(12, 8))
missing_df.head(20)['Percentage'].plot(kind='barh')
plt.xlabel('Percentage Missing')
plt.title('Features with Missing Data')
plt.tight_layout()
plt.show()

### 2.2.1 Missing Data Patterns

Analyzing patterns and relationships in missing data across features.

In [ ]:
# Analyze missing data correlations
import missingno as mnno

# Alternative visualization without missingno
missing_data = df.isnull()
missing_corr = missing_data.corr()

print("Features with correlated missingness:\n")
for i in range(len(missing_corr.columns)):
    for j in range(i+1, len(missing_corr.columns)):
        if abs(missing_corr.iloc[i, j]) > 0.5:
            print(f"{missing_corr.columns[i]} <-> {missing_corr.columns[j]}: {missing_corr.iloc[i, j]:.3f}")

### 2.2.2 Lot Frontage Missing Value Analysis

Lot Frontage has 16.7% missing values. We'll analyze the pattern to inform imputation strategy.

In [ ]:
print("Lot Frontage Missing Value Analysis:\n")
print(f"Total missing: {df['Lot Frontage'].isnull().sum()} ({df['Lot Frontage'].isnull().sum()/len(df)*100:.2f}%)")
print(f"\nLot Frontage statistics (non-missing):")
print(df['Lot Frontage'].describe())

# Check by neighborhood
print("\nLot Frontage by Neighborhood:")
neighborhood_lotfront = df.groupby('Neighborhood')['Lot Frontage'].agg(['count', 'mean', 'median', 'std'])
print(neighborhood_lotfront.head(10))

### 2.3.1 Data Quality and Consistency Checks

Verifying data integrity and identifying potential anomalies.

In [ ]:
# Year-related consistency checks
print("Year-related Data Quality Checks:\n")

# Check for impossible year values
print(f"Year Built range: {df['Year Built'].min()} - {df['Year Built'].max()}")
print(f"Year Remod/Add range: {df['Year Remod/Add'].min()} - {df['Year Remod/Add'].max()}")
print(f"Year Sold range: {df['Yr Sold'].min()} - {df['Yr Sold'].max()}")

# Check for remodel before build
remod_before_build = df[df['Year Remod/Add'] < df['Year Built']]
print(f"\nRecords where remodel year < built year: {len(remod_before_build)}")

# Check garage year
if df['Garage Yr Blt'].notna().sum() > 0:
    garage_before_house = df[df['Garage Yr Blt'] < df['Year Built']].shape[0]
    print(f"Records where garage built before house: {garage_before_house}")

In [ ]:
# Area-related consistency checks
print("Area-related Quality Checks:\n")

# Check if basement area makes sense
print(f"Records with basement area > 1st floor area: {(df['Total Bsmt SF'] > df['1st Flr SF']).sum()}")

# Check total living area consistency
df['Calculated_Liv_Area'] = df['1st Flr SF'] + df['2nd Flr SF'] + df['Low Qual Fin SF']
area_mismatch = df[abs(df['Calculated_Liv_Area'] - df['Gr Liv Area']) > 10]
print(f"Records with living area calculation mismatch: {len(area_mismatch)}")

# Check for zero or negative values
print(f"\nRecords with Lot Area <= 0: {(df['Lot Area'] <= 0).sum()}")
print(f"Records with Gr Liv Area <= 0: {(df['Gr Liv Area'] <= 0).sum()}")
print(f"Records with SalePrice <= 0: {(df['SalePrice'] <= 0).sum()}")

In [ ]:
# Check for duplicate records
print("Duplicate Records Check:\n")
duplicates = df.duplicated(subset=df.columns.difference(['Order']), keep=False)
print(f"Total duplicate rows (excluding Order): {duplicates.sum()}")

if duplicates.sum() > 0:
    print("\nDuplicate records found:")
    print(df[duplicates][['PID', 'SalePrice', 'Gr Liv Area', 'Year Built']].head())
else:
    print("No duplicate records found - data quality check passed.")

### 2.3 Missing Value Treatment Strategy

**Approach:**
1. Drop features with >50% missing (insufficient for imputation)
2. Categorical features: impute with 'None' (indicates absence)
3. Numeric features: impute with median or zero (context-dependent)
4. Lot Frontage: neighborhood-grouped median

In [ ]:
# Drop heavily missing columns
missing_threshold = 0.5
cols_to_drop = missing_pct[missing_pct > missing_threshold].index.tolist()

print(f"Dropping {len(cols_to_drop)} features: {cols_to_drop}")
df_clean = df.drop(columns=cols_to_drop)
print(f"Shape after dropping: {df_clean.shape}")

In [ ]:
# Impute categorical - 'None' means feature doesn't exist
categorical_none = ['Mas Vnr Type', 'Fireplace Qu', 'Garage Type', 'Garage Finish',
                    'Garage Qual', 'Garage Cond', 'Bsmt Qual', 'Bsmt Cond',
                    'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin Type 2']

for col in categorical_none:
    if col in df_clean.columns:
        df_clean[col].fillna('None', inplace=True)

print(f"Imputed {len(categorical_none)} categorical features with 'None'")

In [ ]:
# Impute numeric - 0 for areas/counts where absence means zero
numeric_zero = ['Mas Vnr Area', 'BsmtFin SF 1', 'BsmtFin SF 2', 'Bsmt Unf SF',
                'Total Bsmt SF', 'Bsmt Full Bath', 'Bsmt Half Bath', 
                'Garage Cars', 'Garage Area']

for col in numeric_zero:
    if col in df_clean.columns:
        df_clean[col].fillna(0, inplace=True)

# Lot Frontage: neighborhood-based median
df_clean['Lot Frontage'] = df_clean.groupby('Neighborhood')['Lot Frontage'].transform(
    lambda x: x.fillna(x.median())
)

# Garage Year Built: use house year if no garage
if 'Garage Yr Blt' in df_clean.columns:
    df_clean['Garage Yr Blt'].fillna(df_clean['Year Built'], inplace=True)

# Electrical: mode (only 1 missing)
if df_clean['Electrical'].isnull().sum() > 0:
    df_clean['Electrical'].fillna(df_clean['Electrical'].mode()[0], inplace=True)

print(f"Remaining missing values: {df_clean.isnull().sum().sum()}")

### 2.4 Univariate Analysis - Numerical Features

Examining distributions of numerical features to understand central tendencies, spread, and skewness.

In [ ]:
# Select numeric columns
numeric_cols = df_clean.select_dtypes(include=[np.number]).columns.tolist()
numeric_cols = [col for col in numeric_cols if col not in ['Order', 'PID']]

print(f"Analyzing {len(numeric_cols)} numeric features\n")

# Create histograms for key numeric features
key_features = ['SalePrice', 'Gr Liv Area', 'Lot Area', 'Overall Qual', 
                'Year Built', 'Total Bsmt SF', 'Garage Area']

fig, axes = plt.subplots(3, 3, figsize=(15, 12))
axes = axes.ravel()

for idx, col in enumerate(key_features[:9]):
    if col in df_clean.columns:
        axes[idx].hist(df_clean[col], bins=30, edgecolor='black', alpha=0.7)
        axes[idx].set_title(col)
        axes[idx].set_xlabel('Value')
        axes[idx].set_ylabel('Frequency')

plt.tight_layout()
plt.show()

### 2.5 Target Variable Analysis

Detailed examination of SalePrice distribution.

In [ ]:
print("SalePrice Statistics:")
print(df_clean['SalePrice'].describe())

fig, axes = plt.subplots(1, 2, figsize=(14, 5))

axes[0].hist(df_clean['SalePrice'], bins=50, edgecolor='black')
axes[0].axvline(df_clean['SalePrice'].median(), color='red', linestyle='--', 
                linewidth=2, label=f"Median: ${df_clean['SalePrice'].median():,.0f}")
axes[0].set_xlabel('Sale Price ($)')
axes[0].set_ylabel('Frequency')
axes[0].set_title('Distribution of Sale Prices')
axes[0].legend()

axes[1].boxplot(df_clean['SalePrice'])
axes[1].set_ylabel('Sale Price ($)')
axes[1].set_title('Sale Price Box Plot')

plt.tight_layout()
plt.show()

In [ ]:
# Comprehensive histograms for all numeric features
fig, axes = plt.subplots(6, 5, figsize=(20, 18))
axes = axes.ravel()

numeric_features = ['Lot Area', 'Overall Qual', 'Overall Cond', 'Year Built', 
                    'Year Remod/Add', 'Mas Vnr Area', 'BsmtFin SF 1', 'BsmtFin SF 2',
                    'Bsmt Unf SF', 'Total Bsmt SF', '1st Flr SF', '2nd Flr SF',
                    'Gr Liv Area', 'Bsmt Full Bath', 'Bsmt Half Bath', 'Full Bath',
                    'Half Bath', 'Bedroom AbvGr', 'Kitchen AbvGr', 'TotRms AbvGrd',
                    'Fireplaces', 'Garage Cars', 'Garage Area', 'Wood Deck SF',
                    'Open Porch SF', 'Enclosed Porch', 'Screen Porch', 'Pool Area',
                    'Yr Sold', 'SalePrice']

for idx, col in enumerate(numeric_features[:30]):
    if col in df_clean.columns and idx < 30:
        axes[idx].hist(df_clean[col].dropna(), bins=30, edgecolor='black', alpha=0.7, color='steelblue')
        axes[idx].set_title(f'{col}', fontsize=10, fontweight='bold')
        axes[idx].set_xlabel('')
        axes[idx].set_ylabel('Count', fontsize=8)
        axes[idx].tick_params(labelsize=8)

# Hide unused subplots
for idx in range(len(numeric_features), 30):
    axes[idx].axis('off')

plt.tight_layout()
plt.show()

print(f"Analyzed distributions of {len(numeric_features)} numeric features")

### 2.6 Univariate Analysis - Categorical Features

Examining the distribution and frequency of categorical variables.

In [ ]:
# Select categorical columns
categorical_cols = df_clean.select_dtypes(include=['object']).columns.tolist()
print(f"Analyzing {len(categorical_cols)} categorical features\n")

# Key categorical features for visualization
key_cat = ['MS Zoning', 'Neighborhood', 'Bldg Type', 'House Style', 
           'Foundation', 'Heating QC', 'Central Air', 'Kitchen Qual']

fig, axes = plt.subplots(3, 3, figsize=(18, 12))
axes = axes.ravel()

for idx, col in enumerate(key_cat[:9]):
    if col in df_clean.columns:
        value_counts = df_clean[col].value_counts()
        axes[idx].bar(range(len(value_counts)), value_counts.values, color='steelblue', alpha=0.7)
        axes[idx].set_xticks(range(len(value_counts)))
        axes[idx].set_xticklabels(value_counts.index, rotation=45, ha='right', fontsize=8)
        axes[idx].set_title(f'{col}', fontweight='bold')
        axes[idx].set_ylabel('Frequency')

plt.tight_layout()
plt.show()

In [ ]:
# Additional categorical features
more_cat = ['Roof Style', 'Exterior 1st', 'Exter Qual', 'Bsmt Qual', 
            'Garage Type', 'Garage Finish', 'Paved Drive', 'Sale Condition']

fig, axes = plt.subplots(3, 3, figsize=(18, 12))
axes = axes.ravel()

for idx, col in enumerate(more_cat[:9]):
    if col in df_clean.columns:
        value_counts = df_clean[col].value_counts()
        axes[idx].bar(range(len(value_counts)), value_counts.values, color='coral', alpha=0.7)
        axes[idx].set_xticks(range(len(value_counts)))
        axes[idx].set_xticklabels(value_counts.index, rotation=45, ha='right', fontsize=8)
        axes[idx].set_title(f'{col}', fontweight='bold')
        axes[idx].set_ylabel('Frequency')

plt.tight_layout()
plt.show()

### 2.7 Statistical Summary

Comprehensive statistical overview of numeric features.

In [ ]:
print("Statistical Summary of Key Numeric Features:\n")
key_stats = df_clean[['Lot Area', 'Overall Qual', 'Year Built', 'Total Bsmt SF',
                      'Gr Liv Area', 'Full Bath', 'Garage Area', 'SalePrice']].describe()
print(key_stats)

### 2.8 Correlation Analysis

Identifying relationships between numeric features and the target variable.

In [ ]:
# Calculate correlations with SalePrice
numeric_df = df_clean.select_dtypes(include=[np.number])
correlations = numeric_df.corr()['SalePrice'].sort_values(ascending=False)

print("Top 15 Features Correlated with SalePrice:\n")
print(correlations.head(15))

In [ ]:
# Heatmap of top correlated features
top_features = correlations.head(16).index.tolist()
corr_subset = numeric_df[top_features].corr()

plt.figure(figsize=(12, 10))
sns.heatmap(corr_subset, annot=True, fmt='.2f', cmap='coolwarm', 
            square=True, linewidths=0.5, cbar_kws={"shrink": 0.8})
plt.title('Correlation Heatmap - Top Features', fontsize=14, fontweight='bold')
plt.tight_layout()
plt.show()

### 2.9 Bivariate Analysis - Numeric Features vs SalePrice

Exploring relationships between individual features and sale price.

In [ ]:
# Scatter plots for top correlated features
top_corr_features = ['Overall Qual', 'Gr Liv Area', 'Garage Area', 
                     'Total Bsmt SF', '1st Flr SF', 'Year Built']

fig, axes = plt.subplots(2, 3, figsize=(18, 10))
axes = axes.ravel()

for idx, feature in enumerate(top_corr_features):
    if feature in df_clean.columns:
        axes[idx].scatter(df_clean[feature], df_clean['SalePrice'], 
                         alpha=0.5, s=20, color='steelblue')
        axes[idx].set_xlabel(feature, fontweight='bold')
        axes[idx].set_ylabel('SalePrice', fontweight='bold')
        axes[idx].set_title(f'SalePrice vs {feature}', fontsize=11)
        
        # Add correlation coefficient
        corr_val = df_clean[[feature, 'SalePrice']].corr().iloc[0, 1]
        axes[idx].text(0.05, 0.95, f'r = {corr_val:.3f}', 
                      transform=axes[idx].transAxes, fontsize=10,
                      verticalalignment='top', bbox=dict(boxstyle='round', 
                      facecolor='wheat', alpha=0.5))

plt.tight_layout()
plt.show()

In [ ]:
# Additional scatter plots
more_features = ['Garage Cars', 'Full Bath', 'Year Remod/Add', 
                 'TotRms AbvGrd', 'Fireplaces', 'Lot Area']

fig, axes = plt.subplots(2, 3, figsize=(18, 10))
axes = axes.ravel()

for idx, feature in enumerate(more_features):
    if feature in df_clean.columns:
        axes[idx].scatter(df_clean[feature], df_clean['SalePrice'], 
                         alpha=0.5, s=20, color='coral')
        axes[idx].set_xlabel(feature, fontweight='bold')
        axes[idx].set_ylabel('SalePrice', fontweight='bold')
        axes[idx].set_title(f'SalePrice vs {feature}', fontsize=11)
        
        corr_val = df_clean[[feature, 'SalePrice']].corr().iloc[0, 1]
        axes[idx].text(0.05, 0.95, f'r = {corr_val:.3f}', 
                      transform=axes[idx].transAxes, fontsize=10,
                      verticalalignment='top', bbox=dict(boxstyle='round', 
                      facecolor='wheat', alpha=0.5))

plt.tight_layout()
plt.show()

### 2.10 Bivariate Analysis - Categorical Features vs SalePrice

Examining how categorical variables influence sale price through box plots.

In [ ]:
# Box plots for categorical features vs SalePrice
cat_features = ['Overall Qual', 'Neighborhood', 'Kitchen Qual', 
                'Garage Type', 'Foundation', 'Heating QC']

fig, axes = plt.subplots(2, 3, figsize=(18, 10))
axes = axes.ravel()

for idx, feature in enumerate(cat_features):
    if feature in df_clean.columns:
        # Sort by median for better visualization
        order = df_clean.groupby(feature)['SalePrice'].median().sort_values().index
        
        axes[idx].boxplot([df_clean[df_clean[feature] == cat]['SalePrice'].values 
                          for cat in order],
                         labels=order, patch_artist=True)
        axes[idx].set_xlabel(feature, fontweight='bold')
        axes[idx].set_ylabel('SalePrice', fontweight='bold')
        axes[idx].set_title(f'SalePrice by {feature}', fontsize=11)
        axes[idx].tick_params(axis='x', rotation=45, labelsize=8)

plt.tight_layout()
plt.show()

In [ ]:
# Additional categorical box plots
more_cat_features = ['Bldg Type', 'House Style', 'Central Air', 
                     'Bsmt Qual', 'Exter Qual', 'Sale Condition']

fig, axes = plt.subplots(2, 3, figsize=(18, 10))
axes = axes.ravel()

for idx, feature in enumerate(more_cat_features):
    if feature in df_clean.columns:
        order = df_clean.groupby(feature)['SalePrice'].median().sort_values().index
        
        axes[idx].boxplot([df_clean[df_clean[feature] == cat]['SalePrice'].values 
                          for cat in order],
                         labels=order, patch_artist=True)
        axes[idx].set_xlabel(feature, fontweight='bold')
        axes[idx].set_ylabel('SalePrice', fontweight='bold')
        axes[idx].set_title(f'SalePrice by {feature}', fontsize=11)
        axes[idx].tick_params(axis='x', rotation=45, labelsize=8)

plt.tight_layout()
plt.show()

### 2.11 Outlier Detection

Identifying extreme values using the Interquartile Range (IQR) method.

In [ ]:
def detect_outliers_iqr(data, column):
    """Detect outliers using IQR method"""
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = data[(data[column] < lower_bound) | (data[column] > upper_bound)]
    return outliers, lower_bound, upper_bound

# Check key features for outliers
features_to_check = ['SalePrice', 'Gr Liv Area', 'Lot Area', 'Total Bsmt SF']

print("Outlier Detection Summary:\n")
for feature in features_to_check:
    if feature in df_clean.columns:
        outliers, lower, upper = detect_outliers_iqr(df_clean, feature)
        print(f"{feature}:")
        print(f"  Lower Bound: {lower:.2f}")
        print(f"  Upper Bound: {upper:.2f}")
        print(f"  Number of Outliers: {len(outliers)} ({len(outliers)/len(df_clean)*100:.2f}%)\n")

In [ ]:
# Visualize outliers
fig, axes = plt.subplots(2, 2, figsize=(14, 10))
axes = axes.ravel()

for idx, feature in enumerate(features_to_check):
    if feature in df_clean.columns:
        outliers, lower, upper = detect_outliers_iqr(df_clean, feature)
        
        axes[idx].scatter(df_clean.index, df_clean[feature], 
                         alpha=0.5, s=10, color='steelblue', label='Normal')
        axes[idx].scatter(outliers.index, outliers[feature], 
                         alpha=0.7, s=30, color='red', label='Outlier')
        axes[idx].axhline(y=lower, color='green', linestyle='--', linewidth=1, label='Lower Bound')
        axes[idx].axhline(y=upper, color='orange', linestyle='--', linewidth=1, label='Upper Bound')
        axes[idx].set_xlabel('Index')
        axes[idx].set_ylabel(feature)
        axes[idx].set_title(f'Outliers in {feature}', fontweight='bold')
        axes[idx].legend(fontsize=8)

plt.tight_layout()
plt.show()

### 2.12 Outlier Treatment Decision

After analysis, we retain outliers as they represent legitimate high-value properties and extreme lot sizes that are valid data points in the real estate market.

---\n\n<a id='phase2b'></a>\n## Phase 2B: Feature Engineering\n\n**Objective:** Create new meaningful features to improve model performance.\n\n### 2.13 Engineered Features Creation

In [ ]:
# Create a copy for feature engineering
df_featured = df_clean.copy()

print("Creating engineered features...\n")

# 1. Total Bathrooms
df_featured['Total_Bathrooms'] = (df_featured['Full Bath'] + 
                                   0.5 * df_featured['Half Bath'] + 
                                   df_featured['Bsmt Full Bath'] + 
                                   0.5 * df_featured['Bsmt Half Bath'])
print(f"Created Total_Bathrooms: combines all bathroom counts")

# 2. Total Porch Area
porch_cols = ['Wood Deck SF', 'Open Porch SF', 'Enclosed Porch', 
              '3Ssn Porch', 'Screen Porch']
df_featured['Total_Porch_SF'] = df_featured[porch_cols].sum(axis=1)
print(f"Created Total_Porch_SF: sum of all porch areas")

# 3. House Age
df_featured['House_Age'] = df_featured['Yr Sold'] - df_featured['Year Built']
print(f"Created House_Age: years since construction")

# 4. Years Since Remodel
df_featured['Years_Since_Remod'] = df_featured['Yr Sold'] - df_featured['Year Remod/Add']
print(f"Created Years_Since_Remod: years since last remodel")

# 5. Total Square Footage
df_featured['Total_SF'] = df_featured['Total Bsmt SF'] + df_featured['Gr Liv Area']
print(f"Created Total_SF: total interior square footage")

print(f"\nTotal features now: {df_featured.shape[1]}")
print(f"Original features: {df_clean.shape[1]}")
print(f"New features added: {df_featured.shape[1] - df_clean.shape[1]}")

In [ ]:
# Check correlations of new features with SalePrice
new_features = ['Total_Bathrooms', 'Total_Porch_SF', 'House_Age', 
                'Years_Since_Remod', 'Total_SF']

print("Correlation of Engineered Features with SalePrice:\n")
for feat in new_features:
    corr = df_featured[[feat, 'SalePrice']].corr().iloc[0, 1]
    print(f"{feat:25s}: {corr:.4f}")

In [ ]:
# Visualize engineered features vs SalePrice
fig, axes = plt.subplots(2, 3, figsize=(18, 10))
axes = axes.ravel()

for idx, feature in enumerate(new_features):
    axes[idx].scatter(df_featured[feature], df_featured['SalePrice'], 
                     alpha=0.5, s=20, color='green')
    axes[idx].set_xlabel(feature, fontweight='bold')
    axes[idx].set_ylabel('SalePrice', fontweight='bold')
    axes[idx].set_title(f'SalePrice vs {feature}', fontsize=11)
    
    corr_val = df_featured[[feature, 'SalePrice']].corr().iloc[0, 1]
    axes[idx].text(0.05, 0.95, f'r = {corr_val:.3f}', 
                  transform=axes[idx].transAxes, fontsize=10,
                  verticalalignment='top', bbox=dict(boxstyle='round', 
                  facecolor='lightgreen', alpha=0.5))

axes[5].axis('off')
plt.tight_layout()
plt.show()

### 2.14 Categorical Variable Encoding

Converting categorical variables to numeric format for modeling.

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Create copy for encoding
df_encoded = df_featured.copy()

# Get categorical columns
cat_columns = df_encoded.select_dtypes(include=['object']).columns.tolist()

print(f"Encoding {len(cat_columns)} categorical features...\n")

# Apply label encoding
label_encoders = {}
for col in cat_columns:
    le = LabelEncoder()
    df_encoded[col] = le.fit_transform(df_encoded[col].astype(str))
    label_encoders[col] = le

print(f"Encoding complete. All features are now numeric.")
print(f"\nData shape: {df_encoded.shape}")
print(f"\nData types after encoding:")
print(df_encoded.dtypes.value_counts())

In [ ]:
# Final check for missing values
print("Final missing value check:")
missing_final = df_encoded.isnull().sum().sum()
print(f"Total missing values: {missing_final}")

if missing_final > 0:
    print("\nImputing remaining missing values with median...")
    for col in df_encoded.columns:
        if df_encoded[col].isnull().sum() > 0:
            if df_encoded[col].dtype in ['float64', 'int64']:
                df_encoded[col].fillna(df_encoded[col].median(), inplace=True)
    print(f"After final imputation: {df_encoded.isnull().sum().sum()} missing values")

print(f"\nDataset ready for modeling!")
print(f"Shape: {df_encoded.shape}")

### 2.7.1 Distribution Skewness Analysis

Examining skewness in numerical features to understand transformation needs.

In [ ]:
# Calculate skewness for numeric features
from scipy.stats import skew

numeric_features = df_clean.select_dtypes(include=[np.number]).columns
skewness = df_clean[numeric_features].apply(lambda x: skew(x.dropna()))
skewness_df = pd.DataFrame({'Feature': skewness.index, 'Skewness': skewness.values})
skewness_df = skewness_df.sort_values('Skewness', key=abs, ascending=False)

print("Top 15 Most Skewed Features:\n")
print(skewness_df.head(15).to_string(index=False))

print("\nSkewness Interpretation:")
print("  -0.5 to 0.5: Fairly symmetric")
print("  0.5 to 1.0 or -1.0 to -0.5: Moderately skewed")
print("  >1.0 or <-1.0: Highly skewed")

In [ ]:
# Visualize highly skewed features
highly_skewed = skewness_df[abs(skewness_df['Skewness']) > 2].head(6)['Feature'].tolist()

if len(highly_skewed) > 0:
    fig, axes = plt.subplots(2, 3, figsize=(15, 10))
    axes = axes.ravel()
    
    for idx, feature in enumerate(highly_skewed[:6]):
        if feature in df_clean.columns:
            axes[idx].hist(df_clean[feature].dropna(), bins=30, edgecolor='black', alpha=0.7)
            axes[idx].set_title(f'{feature}\n(Skewness: {skewness[feature]:.2f})', fontweight='bold')
            axes[idx].set_xlabel('Value')
            axes[idx].set_ylabel('Frequency')
    
    for idx in range(len(highly_skewed), 6):
        axes[idx].axis('off')
    
    plt.tight_layout()
    plt.show()

### 2.10.1 Neighborhood Price Analysis

Detailed examination of price variations across neighborhoods.

In [ ]:
# Neighborhood statistics
neighborhood_stats = df_clean.groupby('Neighborhood').agg({
    'SalePrice': ['count', 'mean', 'median', 'std', 'min', 'max']
}).round(2)

neighborhood_stats.columns = ['Count', 'Mean', 'Median', 'Std', 'Min', 'Max']
neighborhood_stats = neighborhood_stats.sort_values('Median', ascending=False)

print("Sale Price Statistics by Neighborhood:\n")
print(neighborhood_stats.head(15))

In [ ]:
# Visualize neighborhood prices
top_neighborhoods = neighborhood_stats.head(10).index.tolist()

fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Bar chart of median prices
median_prices = neighborhood_stats.head(10)['Median'].values
axes[0].barh(range(len(top_neighborhoods)), median_prices, color='steelblue', alpha=0.7)
axes[0].set_yticks(range(len(top_neighborhoods)))
axes[0].set_yticklabels(top_neighborhoods)
axes[0].set_xlabel('Median Sale Price ($)', fontweight='bold')
axes[0].set_title('Top 10 Neighborhoods by Median Price', fontweight='bold')
axes[0].invert_yaxis()

# Box plot of price distribution
neighbor_data = [df_clean[df_clean['Neighborhood'] == n]['SalePrice'].values for n in top_neighborhoods]
axes[1].boxplot(neighbor_data, labels=top_neighborhoods, vert=False)
axes[1].set_xlabel('Sale Price ($)', fontweight='bold')
axes[1].set_title('Price Distribution by Top Neighborhoods', fontweight='bold')

plt.tight_layout()
plt.show()

### 2.10.2 Temporal Analysis - Year Built Impact

Analyzing how construction year affects property values.

In [ ]:
# Year built analysis
print("Year Built Statistics:\n")
print(df_clean['Year Built'].describe())

# Create decade bins
df_clean['Decade_Built'] = (df_clean['Year Built'] // 10) * 10

decade_stats = df_clean.groupby('Decade_Built').agg({
    'SalePrice': ['count', 'mean', 'median'],
    'Gr Liv Area': 'mean'
}).round(2)

print("\nSale Price by Decade Built:\n")
print(decade_stats)

In [ ]:
# Visualize year built impact
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Scatter: Year Built vs SalePrice
axes[0].scatter(df_clean['Year Built'], df_clean['SalePrice'], alpha=0.4, s=20)
axes[0].set_xlabel('Year Built', fontweight='bold')
axes[0].set_ylabel('Sale Price ($)', fontweight='bold')
axes[0].set_title('Sale Price vs Year Built', fontweight='bold')
axes[0].grid(alpha=0.3)

# Bar: Average price by decade
decade_avg = df_clean.groupby('Decade_Built')['SalePrice'].mean()
axes[1].bar(decade_avg.index, decade_avg.values, color='steelblue', alpha=0.7, width=8)
axes[1].set_xlabel('Decade Built', fontweight='bold')
axes[1].set_ylabel('Average Sale Price ($)', fontweight='bold')
axes[1].set_title('Average Sale Price by Decade', fontweight='bold')
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.show()

### 2.10.3 Quality Ratings Impact Analysis

Examining how overall quality and condition ratings affect prices.

In [ ]:
# Quality and condition analysis
print("Overall Quality Distribution:\n")
print(df_clean['Overall Qual'].value_counts().sort_index())

print("\n\nOverall Condition Distribution:\n")
print(df_clean['Overall Cond'].value_counts().sort_index())

# Average price by quality
qual_price = df_clean.groupby('Overall Qual')['SalePrice'].agg(['count', 'mean', 'median'])
print("\n\nSale Price by Overall Quality:\n")
print(qual_price)

In [ ]:
# Visualize quality impact
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Overall Quality vs Price
qual_groups = [df_clean[df_clean['Overall Qual'] == q]['SalePrice'].values for q in sorted(df_clean['Overall Qual'].unique())]
axes[0].boxplot(qual_groups, labels=sorted(df_clean['Overall Qual'].unique()))
axes[0].set_xlabel('Overall Quality Rating', fontweight='bold')
axes[0].set_ylabel('Sale Price ($)', fontweight='bold')
axes[0].set_title('Sale Price Distribution by Overall Quality', fontweight='bold')
axes[0].grid(axis='y', alpha=0.3)

# Overall Condition vs Price
cond_groups = [df_clean[df_clean['Overall Cond'] == c]['SalePrice'].values for c in sorted(df_clean['Overall Cond'].unique())]
axes[1].boxplot(cond_groups, labels=sorted(df_clean['Overall Cond'].unique()))
axes[1].set_xlabel('Overall Condition Rating', fontweight='bold')
axes[1].set_ylabel('Sale Price ($)', fontweight='bold')
axes[1].set_title('Sale Price Distribution by Overall Condition', fontweight='bold')
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.show()

### 2.10.4 Garage Features Analysis

Analyzing the impact of garage characteristics on sale price.

In [ ]:
print("Garage Analysis:\n")
print(f"Properties with garage: {df_clean['Garage Type'].notna().sum()} ({df_clean['Garage Type'].notna().sum()/len(df_clean)*100:.1f}%)")
print(f"Properties without garage: {df_clean['Garage Type'].isna().sum()} ({df_clean['Garage Type'].isna().sum()/len(df_clean)*100:.1f}%)")

print("\n\nGarage Type Distribution:\n")
print(df_clean['Garage Type'].value_counts())

print("\n\nAverage Price by Garage Type:\n")
garage_price = df_clean.groupby('Garage Type')['SalePrice'].agg(['count', 'mean']).sort_values('mean', ascending=False)
print(garage_price)

In [ ]:
# Garage area vs price
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Garage Area vs SalePrice scatter
axes[0].scatter(df_clean['Garage Area'], df_clean['SalePrice'], alpha=0.4, s=20, color='coral')
axes[0].set_xlabel('Garage Area (sq ft)', fontweight='bold')
axes[0].set_ylabel('Sale Price ($)', fontweight='bold')
axes[0].set_title('Sale Price vs Garage Area', fontweight='bold')
axes[0].grid(alpha=0.3)

# Price with/without garage
with_garage = df_clean[df_clean['Garage Area'] > 0]['SalePrice']
without_garage = df_clean[df_clean['Garage Area'] == 0]['SalePrice']

axes[1].boxplot([without_garage, with_garage], labels=['No Garage', 'With Garage'])
axes[1].set_ylabel('Sale Price ($)', fontweight='bold')
axes[1].set_title('Sale Price: With vs Without Garage', fontweight='bold')
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.show()

print(f"\nAverage price with garage: ${with_garage.mean():,.2f}")
print(f"Average price without garage: ${without_garage.mean():,.2f}")
print(f"Price difference: ${with_garage.mean() - without_garage.mean():,.2f}")

### 2.10.5 Basement Features Analysis

Examining basement characteristics and their relationship with price.

In [ ]:
print("Basement Analysis:\n")
print(f"Properties with basement: {(df_clean['Total Bsmt SF'] > 0).sum()} ({(df_clean['Total Bsmt SF'] > 0).sum()/len(df_clean)*100:.1f}%)")
print(f"Properties without basement: {(df_clean['Total Bsmt SF'] == 0).sum()} ({(df_clean['Total Bsmt SF'] == 0).sum()/len(df_clean)*100:.1f}%)")

print("\n\nBasement Quality Distribution:\n")
print(df_clean['Bsmt Qual'].value_counts())

# Average basement size
with_bsmt = df_clean[df_clean['Total Bsmt SF'] > 0]
print(f"\n\nAverage basement size (for homes with basement): {with_bsmt['Total Bsmt SF'].mean():.2f} sq ft")

In [ ]:
# Basement visualizations
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Basement area vs price
axes[0].scatter(df_clean['Total Bsmt SF'], df_clean['SalePrice'], alpha=0.4, s=20, color='purple')
axes[0].set_xlabel('Total Basement Area (sq ft)', fontweight='bold')
axes[0].set_ylabel('Sale Price ($)', fontweight='bold')
axes[0].set_title('Sale Price vs Basement Area', fontweight='bold')
axes[0].grid(alpha=0.3)

# Price by basement quality
bsmt_qual_order = ['None', 'Po', 'Fa', 'TA', 'Gd', 'Ex']
available_quals = [q for q in bsmt_qual_order if q in df_clean['Bsmt Qual'].unique()]
bsmt_qual_groups = [df_clean[df_clean['Bsmt Qual'] == q]['SalePrice'].values for q in available_quals]

axes[1].boxplot(bsmt_qual_groups, labels=available_quals)
axes[1].set_xlabel('Basement Quality', fontweight='bold')
axes[1].set_ylabel('Sale Price ($)', fontweight='bold')
axes[1].set_title('Sale Price by Basement Quality', fontweight='bold')
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.show()

### 2.10.6 Living Area Deep Dive

Detailed analysis of above-grade living area - one of the strongest price predictors.

In [ ]:
print("Living Area (Gr Liv Area) Analysis:\n")
print(df_clean['Gr Liv Area'].describe())

# Create living area categories
df_clean['Liv_Area_Category'] = pd.cut(df_clean['Gr Liv Area'], 
                                        bins=[0, 1000, 1500, 2000, 3000, 6000],
                                        labels=['Very Small', 'Small', 'Medium', 'Large', 'Very Large'])

print("\n\nLiving Area Categories:\n")
print(df_clean['Liv_Area_Category'].value_counts().sort_index())

print("\n\nAverage Price by Living Area Category:\n")
livarea_price = df_clean.groupby('Liv_Area_Category')['SalePrice'].agg(['count', 'mean', 'median'])
print(livarea_price)

In [ ]:
# Living area visualizations
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Histogram with price color coding
scatter = axes[0].scatter(df_clean['Gr Liv Area'], df_clean['SalePrice'], 
                          c=df_clean['SalePrice'], cmap='viridis', alpha=0.6, s=30)
axes[0].set_xlabel('Above Grade Living Area (sq ft)', fontweight='bold')
axes[0].set_ylabel('Sale Price ($)', fontweight='bold')
axes[0].set_title('Sale Price vs Living Area (Color: Price)', fontweight='bold')
plt.colorbar(scatter, ax=axes[0])
axes[0].grid(alpha=0.3)

# Box plot by category
livarea_cat_data = [df_clean[df_clean['Liv_Area_Category'] == cat]['SalePrice'].values 
                    for cat in ['Very Small', 'Small', 'Medium', 'Large', 'Very Large']]
axes[1].boxplot(livarea_cat_data, labels=['Very Small', 'Small', 'Medium', 'Large', 'Very Large'])
axes[1].set_xlabel('Living Area Category', fontweight='bold')
axes[1].set_ylabel('Sale Price ($)', fontweight='bold')
axes[1].set_title('Price Distribution by Living Area Category', fontweight='bold')
axes[1].tick_params(axis='x', rotation=15)
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.show()

### 2.10.7 Lot Size Analysis

Investigating the relationship between lot size and property values.

In [ ]:
print("Lot Area Analysis:\n")
print(df_clean['Lot Area'].describe())

# Create lot size categories
df_clean['Lot_Size_Category'] = pd.cut(df_clean['Lot Area'], 
                                       bins=[0, 5000, 10000, 15000, 20000, 200000],
                                       labels=['Small', 'Medium', 'Large', 'Very Large', 'Exceptional'])

print("\n\nLot Size Distribution:\n")
print(df_clean['Lot_Size_Category'].value_counts().sort_index())

lot_price_stats = df_clean.groupby('Lot_Size_Category')['SalePrice'].agg(['count', 'mean', 'median'])
print("\n\nAverage Price by Lot Size:\n")
print(lot_price_stats)

In [ ]:
# Lot size visualizations
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Scatter plot (with outlier handling for visualization)
lot_plot_data = df_clean[df_clean['Lot Area'] < 30000]  # Filter extreme outliers for better viz
axes[0].scatter(lot_plot_data['Lot Area'], lot_plot_data['SalePrice'], alpha=0.4, s=20, color='green')
axes[0].set_xlabel('Lot Area (sq ft)', fontweight='bold')
axes[0].set_ylabel('Sale Price ($)', fontweight='bold')
axes[0].set_title('Sale Price vs Lot Area (filtered <30K sq ft)', fontweight='bold')
axes[0].grid(alpha=0.3)

# Bar chart by category
lot_avg_price = df_clean.groupby('Lot_Size_Category')['SalePrice'].mean()
axes[1].bar(range(len(lot_avg_price)), lot_avg_price.values, color='green', alpha=0.7)
axes[1].set_xticks(range(len(lot_avg_price)))
axes[1].set_xticklabels(lot_avg_price.index, rotation=15)
axes[1].set_ylabel('Average Sale Price ($)', fontweight='bold')
axes[1].set_xlabel('Lot Size Category', fontweight='bold')
axes[1].set_title('Average Price by Lot Size Category', fontweight='bold')
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.show()

### 2.10.8 Building Type and Style Analysis

Analyzing how building type and architectural style affect pricing.

In [ ]:
print("Building Type Analysis:\n")
print(df_clean['Bldg Type'].value_counts())

print("\n\nAverage Price by Building Type:\n")
bldg_price = df_clean.groupby('Bldg Type')['SalePrice'].agg(['count', 'mean', 'median']).sort_values('mean', ascending=False)
print(bldg_price)

print("\n\nHouse Style Distribution:\n")
print(df_clean['House Style'].value_counts())

print("\n\nAverage Price by House Style:\n")
style_price = df_clean.groupby('House Style')['SalePrice'].agg(['count', 'mean', 'median']).sort_values('mean', ascending=False)
print(style_price.head(10))

In [ ]:
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Building type box plot
bldg_types = df_clean['Bldg Type'].unique()
bldg_data = [df_clean[df_clean['Bldg Type'] == bt]['SalePrice'].values for bt in bldg_types]
axes[0].boxplot(bldg_data, labels=bldg_types)
axes[0].set_xlabel('Building Type', fontweight='bold')
axes[0].set_ylabel('Sale Price ($)', fontweight='bold')
axes[0].set_title('Sale Price by Building Type', fontweight='bold')
axes[0].tick_params(axis='x', rotation=20)
axes[0].grid(axis='y', alpha=0.3)

# House style - top styles only
top_styles = df_clean['House Style'].value_counts().head(8).index
style_data = [df_clean[df_clean['House Style'] == hs]['SalePrice'].values for hs in top_styles]
axes[1].boxplot(style_data, labels=top_styles)
axes[1].set_xlabel('House Style', fontweight='bold')
axes[1].set_ylabel('Sale Price ($)', fontweight='bold')
axes[1].set_title('Sale Price by House Style (Top 8)', fontweight='bold')
axes[1].tick_params(axis='x', rotation=20)
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.show()

### 2.10.9 Sale Characteristics Analysis

Examining sale conditions and timing effects on price.

In [ ]:
print("Sale Condition Analysis:\n")
print(df_clean['Sale Condition'].value_counts())

print("\n\nAverage Price by Sale Condition:\n")
sale_cond_price = df_clean.groupby('Sale Condition')['SalePrice'].agg(['count', 'mean', 'median']).sort_values('mean', ascending=False)
print(sale_cond_price)

print("\n\nSales by Month:\n")
month_sales = df_clean['Mo Sold'].value_counts().sort_index()
print(month_sales)

print("\n\nAverage Price by Month:\n")
month_price = df_clean.groupby('Mo Sold')['SalePrice'].mean().sort_index()
print(month_price)

In [ ]:
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Sales by month
month_labels = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
sales_count = df_clean['Mo Sold'].value_counts().sort_index()
axes[0].bar(range(1, 13), [sales_count.get(i, 0) for i in range(1, 13)], color='steelblue', alpha=0.7)
axes[0].set_xticks(range(1, 13))
axes[0].set_xticklabels(month_labels, rotation=45)
axes[0].set_ylabel('Number of Sales', fontweight='bold')
axes[0].set_xlabel('Month', fontweight='bold')
axes[0].set_title('Sales Volume by Month', fontweight='bold')
axes[0].grid(axis='y', alpha=0.3)

# Average price by month
avg_price_month = df_clean.groupby('Mo Sold')['SalePrice'].mean().sort_index()
axes[1].plot(range(1, 13), [avg_price_month.get(i, 0) for i in range(1, 13)], 
            marker='o', linewidth=2, markersize=8, color='coral')
axes[1].set_xticks(range(1, 13))
axes[1].set_xticklabels(month_labels, rotation=45)
axes[1].set_ylabel('Average Sale Price ($)', fontweight='bold')
axes[1].set_xlabel('Month', fontweight='bold')
axes[1].set_title('Average Sale Price by Month', fontweight='bold')
axes[1].grid(alpha=0.3)

plt.tight_layout()
plt.show()

### 2.11 Advanced Correlation Analysis

Deeper investigation into feature correlations and multicollinearity.

In [ ]:
# Find highly correlated feature pairs (potential multicollinearity)
print("Highly Correlated Feature Pairs (|r| > 0.8):\n")

numeric_corr = df_clean.select_dtypes(include=[np.number]).corr()
high_corr_pairs = []

for i in range(len(numeric_corr.columns)):
    for j in range(i+1, len(numeric_corr.columns)):
        if abs(numeric_corr.iloc[i, j]) > 0.8:
            high_corr_pairs.append((
                numeric_corr.columns[i], 
                numeric_corr.columns[j], 
                numeric_corr.iloc[i, j]
            ))

high_corr_df = pd.DataFrame(high_corr_pairs, columns=['Feature 1', 'Feature 2', 'Correlation'])
high_corr_df = high_corr_df.sort_values('Correlation', key=abs, ascending=False)
print(high_corr_df.to_string(index=False))

### 2.15 Pre-Modeling Data Summary

Final overview of cleaned and engineered dataset before model training.

In [ ]:
print("="*80)
print("FINAL DATASET SUMMARY - READY FOR MODELING")
print("="*80)
print(f"\nTotal Records: {df_encoded.shape[0]:,}")
print(f"Total Features: {df_encoded.shape[1]}")
print(f"  - Numeric Features: {len(df_encoded.select_dtypes(include=[np.number]).columns)}")
print(f"  - Categorical Features (encoded): {len(df_encoded.select_dtypes(include=[np.number]).columns) - len(df.select_dtypes(include=[np.number]).columns)}")
print(f"\nMissing Values: {df_encoded.isnull().sum().sum()}")
print(f"Duplicate Records: {df_encoded.duplicated().sum()}")
print(f"\nTarget Variable (SalePrice):")
print(f"  - Mean: ${df_encoded['SalePrice'].mean():,.2f}")
print(f"  - Median: ${df_encoded['SalePrice'].median():,.2f}")
print(f"  - Min: ${df_encoded['SalePrice'].min():,.2f}")
print(f"  - Max: ${df_encoded['SalePrice'].max():,.2f}")
print(f"  - Std Dev: ${df_encoded['SalePrice'].std():,.2f}")
print(f"\nEngineered Features Added: 5")
print("  - Total_Bathrooms")
print("  - Total_Porch_SF")
print("  - House_Age")
print("  - Years_Since_Remod")
print("  - Total_SF")
print("\nData Quality: All checks passed")
print("Status: Ready for machine learning model training")
print("="*80)

---\n\n<a id='phase3'></a>\n## Phase 3: Model Development & Evaluation\n\n**Objective:** Build and evaluate regression models to predict sale prices.\n\n### 3.1 Data Preparation for Modeling

In [ ]:
# Separate features and target
X = df_encoded.drop(['SalePrice', 'Order', 'PID'], axis=1, errors='ignore')
y = df_encoded['SalePrice']

print(f"Features shape: {X.shape}")
print(f"Target shape: {y.shape}")
print(f"\nNumber of features for modeling: {X.shape[1]}")

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
import math

print("Machine learning libraries imported successfully.")

In [ ]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("Data split completed:")
print(f"Training set: {X_train.shape[0]} samples ({X_train.shape[0]/len(X)*100:.1f}%)")
print(f"Testing set: {X_test.shape[0]} samples ({X_test.shape[0]/len(X)*100:.1f}%)")
print(f"\nTraining features shape: {X_train.shape}")
print(f"Testing features shape: {X_test.shape}")

### 3.2 Model 1: Simple Linear Regression\n\nBuilding a baseline model using the single most correlated feature.

In [ ]:
# Find the feature with highest correlation to SalePrice
correlations = df_encoded.drop(['SalePrice', 'Order', 'PID'], axis=1, errors='ignore').corrwith(df_encoded['SalePrice']).abs()
best_feature = correlations.idxmax()

print(f"Best single feature for prediction: {best_feature}")
print(f"Correlation with SalePrice: {correlations[best_feature]:.4f}")

# Prepare data for simple model
X_train_simple = X_train[[best_feature]]
X_test_simple = X_test[[best_feature]]

In [ ]:
# Train Simple Linear Regression
model_simple = LinearRegression()
model_simple.fit(X_train_simple, y_train)

print("Simple Linear Regression Model Trained")
print(f"\nModel Coefficient: {model_simple.coef_[0]:.2f}")
print(f"Model Intercept: {model_simple.intercept_:.2f}")

# Make predictions
y_train_pred_simple = model_simple.predict(X_train_simple)
y_test_pred_simple = model_simple.predict(X_test_simple)

In [ ]:
# Evaluate Simple Linear Regression
r2_train_simple = r2_score(y_train, y_train_pred_simple)
r2_test_simple = r2_score(y_test, y_test_pred_simple)
rmse_test_simple = math.sqrt(mean_squared_error(y_test, y_test_pred_simple))
mae_test_simple = mean_absolute_error(y_test, y_test_pred_simple)

print("="*60)
print("SIMPLE LINEAR REGRESSION - RESULTS")
print("="*60)
print(f"Feature Used: {best_feature}")
print(f"\nTraining Performance:")
print(f"  R² Score: {r2_train_simple:.4f}")
print(f"\nTesting Performance:")
print(f"  R² Score: {r2_test_simple:.4f}")
print(f"  RMSE: ${rmse_test_simple:,.2f}")
print(f"  MAE: ${mae_test_simple:,.2f}")
print("="*60)

In [ ]:
# Visualize Simple Linear Regression results
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Scatter plot with regression line
axes[0].scatter(X_test_simple, y_test, alpha=0.5, s=30, label='Actual', color='steelblue')
axes[0].plot(X_test_simple, y_test_pred_simple, color='red', linewidth=2, label='Predicted')
axes[0].set_xlabel(best_feature, fontsize=12, fontweight='bold')
axes[0].set_ylabel('SalePrice', fontsize=12, fontweight='bold')
axes[0].set_title(f'Simple Linear Regression: {best_feature} vs SalePrice', fontsize=13, fontweight='bold')
axes[0].legend()
axes[0].grid(alpha=0.3)

# Actual vs Predicted
axes[1].scatter(y_test, y_test_pred_simple, alpha=0.5, s=30, color='steelblue')
axes[1].plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], 
             'r--', linewidth=2, label='Perfect Prediction')
axes[1].set_xlabel('Actual SalePrice', fontsize=12, fontweight='bold')
axes[1].set_ylabel('Predicted SalePrice', fontsize=12, fontweight='bold')
axes[1].set_title(f'Actual vs Predicted (R² = {r2_test_simple:.4f})', fontsize=13, fontweight='bold')
axes[1].legend()
axes[1].grid(alpha=0.3)

plt.tight_layout()
plt.show()

In [ ]:
# Residual analysis for simple model
residuals_simple = y_test - y_test_pred_simple

fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Residual plot
axes[0].scatter(y_test_pred_simple, residuals_simple, alpha=0.5, s=30, color='steelblue')
axes[0].axhline(y=0, color='red', linestyle='--', linewidth=2)
axes[0].set_xlabel('Predicted SalePrice', fontsize=12, fontweight='bold')
axes[0].set_ylabel('Residuals', fontsize=12, fontweight='bold')
axes[0].set_title('Residual Plot - Simple Linear Regression', fontsize=13, fontweight='bold')
axes[0].grid(alpha=0.3)

# Residual distribution
axes[1].hist(residuals_simple, bins=30, edgecolor='black', alpha=0.7, color='steelblue')
axes[1].set_xlabel('Residuals', fontsize=12, fontweight='bold')
axes[1].set_ylabel('Frequency', fontsize=12, fontweight='bold')
axes[1].set_title('Distribution of Residuals', fontsize=13, fontweight='bold')
axes[1].axvline(x=0, color='red', linestyle='--', linewidth=2)
axes[1].grid(alpha=0.3)

plt.tight_layout()
plt.show()

print(f"Mean Residual: ${residuals_simple.mean():,.2f}")
print(f"Std Residual: ${residuals_simple.std():,.2f}")

### 3.3 Model 2: Multiple Linear Regression\n\nBuilding a comprehensive model using all available features.

In [ ]:
# Train Multiple Linear Regression
model_multiple = LinearRegression()
model_multiple.fit(X_train, y_train)

print("Multiple Linear Regression Model Trained")
print(f"\nNumber of features used: {X_train.shape[1]}")
print(f"Model Intercept: {model_multiple.intercept_:.2f}")
print(f"\nTop 10 Feature Coefficients:")

# Get feature coefficients
coef_df = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': model_multiple.coef_
}).sort_values('Coefficient', key=abs, ascending=False)

print(coef_df.head(10))

# Make predictions
y_train_pred_multiple = model_multiple.predict(X_train)
y_test_pred_multiple = model_multiple.predict(X_test)

In [ ]:
# Evaluate Multiple Linear Regression
r2_train_multiple = r2_score(y_train, y_train_pred_multiple)
r2_test_multiple = r2_score(y_test, y_test_pred_multiple)
rmse_test_multiple = math.sqrt(mean_squared_error(y_test, y_test_pred_multiple))
mae_test_multiple = mean_absolute_error(y_test, y_test_pred_multiple)

print("="*60)
print("MULTIPLE LINEAR REGRESSION - RESULTS")
print("="*60)
print(f"Features Used: {X_train.shape[1]} features")
print(f"\nTraining Performance:")
print(f"  R² Score: {r2_train_multiple:.4f}")
print(f"\nTesting Performance:")
print(f"  R² Score: {r2_test_multiple:.4f}")
print(f"  RMSE: ${rmse_test_multiple:,.2f}")
print(f"  MAE: ${mae_test_multiple:,.2f}")
print("="*60)

In [ ]:
# Visualize Multiple Linear Regression results
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Actual vs Predicted
axes[0].scatter(y_test, y_test_pred_multiple, alpha=0.5, s=30, color='green')
axes[0].plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], 
             'r--', linewidth=2, label='Perfect Prediction')
axes[0].set_xlabel('Actual SalePrice', fontsize=12, fontweight='bold')
axes[0].set_ylabel('Predicted SalePrice', fontsize=12, fontweight='bold')
axes[0].set_title(f'Actual vs Predicted (R² = {r2_test_multiple:.4f})', fontsize=13, fontweight='bold')
axes[0].legend()
axes[0].grid(alpha=0.3)

# Prediction error distribution
errors = y_test - y_test_pred_multiple
axes[1].hist(errors, bins=30, edgecolor='black', alpha=0.7, color='green')
axes[1].set_xlabel('Prediction Error', fontsize=12, fontweight='bold')
axes[1].set_ylabel('Frequency', fontsize=12, fontweight='bold')
axes[1].set_title('Distribution of Prediction Errors', fontsize=13, fontweight='bold')
axes[1].axvline(x=0, color='red', linestyle='--', linewidth=2)
axes[1].grid(alpha=0.3)

plt.tight_layout()
plt.show()

In [ ]:
# Residual analysis for multiple model
residuals_multiple = y_test - y_test_pred_multiple

fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Residual plot
axes[0].scatter(y_test_pred_multiple, residuals_multiple, alpha=0.5, s=30, color='green')
axes[0].axhline(y=0, color='red', linestyle='--', linewidth=2)
axes[0].set_xlabel('Predicted SalePrice', fontsize=12, fontweight='bold')
axes[0].set_ylabel('Residuals', fontsize=12, fontweight='bold')
axes[0].set_title('Residual Plot - Multiple Linear Regression', fontsize=13, fontweight='bold')
axes[0].grid(alpha=0.3)

# Q-Q plot approximation
from scipy import stats
stats.probplot(residuals_multiple, dist="norm", plot=axes[1])
axes[1].set_title('Q-Q Plot - Residuals Normality Check', fontsize=13, fontweight='bold')
axes[1].grid(alpha=0.3)

plt.tight_layout()
plt.show()

print(f"Mean Residual: ${residuals_multiple.mean():,.2f}")
print(f"Std Residual: ${residuals_multiple.std():,.2f}")

### 3.4 Feature Importance Analysis\n\nIdentifying which features contribute most to price predictions.

In [ ]:
# Analyze feature importance from coefficients
print("Top 15 Most Important Features (by absolute coefficient):")
print("\n", coef_df.head(15).to_string(index=False))

# Visualize top 20 features
top_20 = coef_df.head(20)

plt.figure(figsize=(12, 8))
plt.barh(range(len(top_20)), top_20['Coefficient'].values, color='steelblue', alpha=0.7)
plt.yticks(range(len(top_20)), top_20['Feature'].values)
plt.xlabel('Coefficient Value', fontsize=12, fontweight='bold')
plt.title('Top 20 Feature Coefficients in Multiple Linear Regression', 
          fontsize=14, fontweight='bold')
plt.axvline(x=0, color='red', linestyle='--', linewidth=1)
plt.grid(axis='x', alpha=0.3)
plt.tight_layout()
plt.show()

### 3.5 Model Comparison\n\nComparing the performance of both models to determine the best approach.

In [ ]:
# Create comparison dataframe
comparison = pd.DataFrame({
    'Metric': ['Features Used', 'R² (Training)', 'R² (Testing)', 'RMSE', 'MAE'],
    'Simple Linear Regression': [
        f'1 ({best_feature})',
        f'{r2_train_simple:.4f}',
        f'{r2_test_simple:.4f}',
        f'${rmse_test_simple:,.2f}',
        f'${mae_test_simple:,.2f}'
    ],
    'Multiple Linear Regression': [
        f'{X_train.shape[1]} features',
        f'{r2_train_multiple:.4f}',
        f'{r2_test_multiple:.4f}',
        f'${rmse_test_multiple:,.2f}',
        f'${mae_test_multiple:,.2f}'
    ]
})

print("\n" + "="*80)
print("MODEL PERFORMANCE COMPARISON")
print("="*80)
print(comparison.to_string(index=False))
print("="*80)

In [ ]:
# Visualize model comparison
metrics = ['R² Score (Test)', 'RMSE', 'MAE']
simple_scores = [r2_test_simple, rmse_test_simple, mae_test_simple]
multiple_scores = [r2_test_multiple, rmse_test_multiple, mae_test_multiple]

fig, axes = plt.subplots(1, 3, figsize=(18, 5))

# R² comparison
axes[0].bar(['Simple LR', 'Multiple LR'], 
            [r2_test_simple, r2_test_multiple], 
            color=['steelblue', 'green'], alpha=0.7)
axes[0].set_ylabel('R² Score', fontsize=11, fontweight='bold')
axes[0].set_title('R² Score Comparison', fontsize=12, fontweight='bold')
axes[0].set_ylim([0, 1])
for i, v in enumerate([r2_test_simple, r2_test_multiple]):
    axes[0].text(i, v + 0.02, f'{v:.4f}', ha='center', fontweight='bold')

# RMSE comparison
axes[1].bar(['Simple LR', 'Multiple LR'], 
            [rmse_test_simple, rmse_test_multiple], 
            color=['steelblue', 'green'], alpha=0.7)
axes[1].set_ylabel('RMSE ($)', fontsize=11, fontweight='bold')
axes[1].set_title('RMSE Comparison (Lower is Better)', fontsize=12, fontweight='bold')
for i, v in enumerate([rmse_test_simple, rmse_test_multiple]):
    axes[1].text(i, v + 1000, f'${v:,.0f}', ha='center', fontweight='bold')

# MAE comparison
axes[2].bar(['Simple LR', 'Multiple LR'], 
            [mae_test_simple, mae_test_multiple], 
            color=['steelblue', 'green'], alpha=0.7)
axes[2].set_ylabel('MAE ($)', fontsize=11, fontweight='bold')
axes[2].set_title('MAE Comparison (Lower is Better)', fontsize=12, fontweight='bold')
for i, v in enumerate([mae_test_simple, mae_test_multiple]):
    axes[2].text(i, v + 1000, f'${v:,.0f}', ha='center', fontweight='bold')

plt.tight_layout()
plt.show()

### 3.6 Sample Predictions\n\nExamining predictions on specific test examples.

In [ ]:
# Show sample predictions
sample_size = 10
sample_indices = np.random.choice(X_test.index, sample_size, replace=False)

prediction_df = pd.DataFrame({
    'Actual Price': y_test.loc[sample_indices].values,
    'Simple LR Prediction': model_simple.predict(X_test.loc[sample_indices, [best_feature]]),
    'Multiple LR Prediction': model_multiple.predict(X_test.loc[sample_indices]),
})

prediction_df['Simple LR Error'] = prediction_df['Actual Price'] - prediction_df['Simple LR Prediction']
prediction_df['Multiple LR Error'] = prediction_df['Actual Price'] - prediction_df['Multiple LR Prediction']

# Format as currency
for col in prediction_df.columns:
    if col != 'Index':
        prediction_df[col] = prediction_df[col].apply(lambda x: f'${x:,.0f}')

print("Sample Predictions from Test Set:\n")
print(prediction_df.to_string(index=False))

### 3.7 Conclusions and Insights

## Key Findings:\n\n### Model Performance\n\n**Simple Linear Regression:**\n- Uses the most correlated single feature as predictor\n- Provides interpretable baseline results\n- Limited by single feature constraint\n\n**Multiple Linear Regression:**\n- Leverages all engineered and original features\n- Achieves substantially higher R² score\n- Demonstrates significant improvement in prediction accuracy\n\n### Feature Engineering Impact\n\nThe engineered features (Total_Bathrooms, Total_Porch_SF, House_Age, Years_Since_Remod, Total_SF) showed strong correlations with sale price and contributed meaningfully to model performance.\n\n### Model Selection\n\nThe Multiple Linear Regression model is recommended for deployment due to:\n1. Significantly higher R² score indicating better fit\n2. Lower prediction errors across RMSE and MAE metrics\n3. Better ability to capture complex relationships between features\n4. Comprehensive use of available property information\n\n### Business Application\n\nThe final model provides reliable price predictions suitable for:\n- Initial property valuations and appraisals\n- Market trend analysis and forecasting\n- Investment decision support\n- Real estate portfolio management\n\n### Future Improvements\n\n1. **Advanced Models:** Explore Random Forest, Gradient Boosting, or Neural Networks for non-linear relationships\n2. **Feature Selection:** Apply techniques like LASSO or Recursive Feature Elimination to reduce dimensionality\n3. **Cross-Validation:** Implement k-fold CV for more robust performance estimates\n4. **Hyperparameter Tuning:** Optimize model parameters systematically using GridSearchCV\n5. **Ensemble Methods:** Combine multiple models for improved predictions\n6. **Feature Interactions:** Create polynomial features to capture interactions\n7. **Regularization:** Apply Ridge or LASSO regression to handle multicollinearity

In [ ]:
# Final summary
print("\n" + "="*80)
print("PROJECT COMPLETION SUMMARY")
print("="*80)
print(f"Dataset: Ames Housing (2,930 properties)")
print(f"Original Features: 82")
print(f"Features after Engineering: {df_featured.shape[1]}")
print(f"Features used in Final Model: {X.shape[1]}")
print(f"\nBest Model: Multiple Linear Regression")
print(f"Final Test R²: {r2_test_multiple:.4f}")
print(f"Final Test RMSE: ${rmse_test_multiple:,.2f}")
print(f"Final Test MAE: ${mae_test_multiple:,.2f}")
print(f"\nModel explains {r2_test_multiple*100:.2f}% of variance in sale prices.")
print("="*80)